src/backend/utils/misc/queryenvironment.c

```diff
@@ -241,6 +246,83 @@ bool has_existing_enr_relations()
 	return false;
 }
 
+/* This only stores catcache inval messages meant for ENR tables in the current transaction. */
+void SaveCatcacheMessage(int cacheId,
+							 uint32 hashValue,
+							 Oid dbId)
+{
+	SharedInvalCatcacheMsg *msg = (SharedInvalCatcacheMsg *) palloc0(sizeof(SharedInvalCatcacheMsg));
+	msg->id = cacheId;
+	msg->dbId = dbId;
+	msg->hashValue = hashValue;
+
+	currentQueryEnv->savedCatcacheMessages = lappend(currentQueryEnv->savedCatcacheMessages, msg);
+}
+
+/* Clear any saved catcache messages at end of xact. */
+void ClearSavedCatcacheMessages()
+{
+	if (!currentQueryEnv)
+		return;
+
+	list_free_deep(currentQueryEnv->savedCatcacheMessages);
+	currentQueryEnv->savedCatcacheMessages = NIL;
+}
+
+/*
+ * SIMessageIsForTempTable
+ * 
+ * Determine whether the msg sent is for a temp table. 
+ * TSQL style temp tables do not need to add messages to the
+ * SI queue, as catalog changes are all session-local.
+ * 
+ * See LocalExecuteInvalidationMessage
+ */
+bool SIMessageIsForTempTable(const SharedInvalidationMessage *msg)
+{
+	if (sql_dialect != SQL_DIALECT_TSQL || temp_oid_buffer_size == 0)
+		return false;
+
+	if (msg->id >= 0)
```

Comment:
this cannot be a switch, because msg->id is not a constant: 

```
queryenvironment.c:309:3: error: case label does not reduce to an integer constant
```